source/slang/slang-check-decl.cpp

```diff
@@ -3652,19 +3682,19 @@ namespace Slang
         checkCallableDeclCommon(decl);
     }
 
-    void SemanticsDeclHeaderVisitor::visitSubscriptDecl(SubscriptDecl* decl)
+    void SemanticsDeclHeaderVisitor::visitAbstractStorageDeclCommon(ContainerDecl* decl)
     {
-        decl->returnType = CheckUsableType(decl->returnType);
-
-        // If we have a subscript declaration with no accessor declarations,
+        // If we have a subscript or property declaration with no accessor declarations,
         // then we should create a single `GetterDecl` to represent
         // the implicit meaning of their declaration, so:
         //
         //      subscript(uint index) -> T;
```

Comment:
A niggle: The use of the colon is presumably to try and match a 'modern var' variable : type? It seems a little inconsistent with the use of -> on subscript, in light of the getter/setting being 'functions'. That some languages use : instead of -> also adds to the worry.

Reply:
I think we should avoid mixing different syntax styles. I agree that we should also support the same C# property syntax `property Type Name`. When you are actively developing using C++/C#, switching to another syntax is a mental overhead and reduces producitivity.

Reply:
There's a few distinct points being discussed here, so I want to tackle each separately:

### On using `:` vs `->`

In the world of "modern" languages,  the form `x : Y` is usually used to indicate that the type of `x` is `Y`. That is consistent with how it is being used here. `property x : Y` is saying that the type of property `x` is `Y`.

The `->` operator is used to construct function types, so that `Y -> Z` is the type of a function from `Y` to `Z`. If we had a variable of function type, we might say `f : Y -> Z`, and that all Just Works.

Many languages (including Swift, where we get much of our inspiration for new syntax) use `->` as the way to introduce the return type of a function:  e.g.,`func f(x: Y) -> Z`. In this case the use of `->` is to reinforce that the type of a function declared this way matches the type of a function-typed variable. E.g.:

```
// These two declarations both have function types, and the types those functions return is `Z`.
func f(x : Y) -> Z { ... }
let g : Y -> Z = ...
```

Using a colon would then be misleading:

```
func f(x : Y) : Z { ... }
let g : Y -> Z = ...
```

This second option starts to look like we are saying `f` has type `Z`, when that isn't the case. You could argue that it should be read as saying that `f(y)` has type `Z` (that is, once we apply the function the type is `Z`), but at the very least we should acknowledge that each choice has arguments in its favor.

### On using `name : Type` vs. `Type name`

The mental overhead to switch styles is a valid point. We obviously need to support the legacy C-like syntax in Slang for HLSL compatibility, so if we were only going to support one syntax that would have to be the one. There is an argument to be made that having two different declaration syntaxes co-exist in Slang is a Bad Thing.

All that said, C/C++-style declaration syntax is a train wreck, and there are things you can't do within its constraints. Even C++ ran into this problem and had to introduce a trailing return-type syntax as a band-aid.

My opinion has been that Slang should provide a slow and gentle ramp from current HLSL syntax/style toward something cleaner and more modern. Whenever a 100% new piece of syntax is being added and exposed to users, I've been trying to expose it in the way that we'd like it to look and not in the way that most closely mimics the legacy of HLSL.

It seems like we should probably have a discussion about the goals here, and arrive at a consensus about the design principles for how to evolve Slang forward. So far I've added things like `let`, `var`, and `func` declarations that all use this "modern" declaration style. If we want to change things now we will have to revisit those designs and either remove them or find a way to align them with whatever design we pick.